cli/ops/websocket.rs

```diff
@@ -0,0 +1,221 @@
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: SendArgs = serde_json::from_value(args)?;
+  let mut resource_table = isolate_state.resource_table.borrow_mut();
+  let stream = resource_table
+    .get_mut::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+    .ok_or_else(OpError::bad_resource_id)?;
+  block_on(stream.send(match args.text {
+    Some(text) => Message::Text(text),
+    None => Message::Binary(zero_copy[0].to_owned().to_vec()),
+  }))
+  .unwrap();
+  Ok(JsonOp::Sync(json!({})))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CloseArgs {
+  rid: u32,
+  code: Option<u16>,
+  reason: Option<String>,
+}
+
+pub fn op_ws_close(
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: CloseArgs = serde_json::from_value(args)?;
+  let resource_table = isolate_state.resource_table.clone();
+  let future = async move {
+    let mut stream = {
+      let mut resource_table = resource_table.borrow_mut();
+      resource_table
+        .remove::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+        .ok_or_else(OpError::bad_resource_id)?
+    };
+    (*stream)
+      .close(Some(CloseFrame {
+        code: CloseCode::from(args.code.unwrap_or(1005)),
+        reason: match args.reason {
+          Some(reason) => Cow::from(reason),
+          None => Default::default(),
+        },
+      }))
+      .await
+      .unwrap();
+
+    Ok(json!({}))
+  };
+
+  Ok(JsonOp::Async(future.boxed_local()))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct NextEventArgs {
+  rid: u32,
+}
+
+pub fn op_ws_next_event(
+  isolate_state: &mut CoreIsolateState,
+  _state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let args: NextEventArgs = serde_json::from_value(args)?;
+  let resource_table = isolate_state.resource_table.clone();
+  let future = async move {
+    let mut resource_table = resource_table.borrow_mut();
+    let stream = resource_table
+      .get_mut::<WebSocketStream<MaybeTlsStream<TcpStream>>>(args.rid)
+      .ok_or_else(OpError::bad_resource_id)?;
+    let message = match stream.next().await.unwrap() {
+      Ok(message) => message,
+      Err(e) => match e {
+        Error::ConnectionClosed => {
+          resource_table
+            .close(args.rid)
+            .ok_or_else(OpError::bad_resource_id)?;
+          return Ok(json!({
+            "type": "closed",
+          }));
+        }
+        _ => {
+          return Ok(json!({
+            "type": "error",
+          }));
+        }
+      },
+    };
+
+    match message {
+      Message::Text(text) => Ok(json!({
+        "type": "string",
+        "data": text
+      })),
+      Message::Binary(data) => {
+        Ok(json!({ //TODO: don't use json to send binary data
```

Comment:
Why not? I think that's fine

Reply:
A WS message is limited at 9.22 exabytes. sure, no one is gonna reach that far, but even a few gigs, and that over json? not sure if that is great

Reply:
I think he means it's fine for now. Of course in the limit we want some sort of zero-copy.